doc/man/nsswitch-config.8
.TH NSSWITCH-CONFIG "8" "23 Oct 2025" "nsswitch-config" "managing nsswitch.conf"
.SH NAME
nsswitch.conf - 
.SH SYNOPSIS
nsswitch-config [OPTIONS]
.SH DESCRIPTION
Tool which merges different nsswitch configuration file snippets into one single
nsswitch.conf file. This file is located in the /etc directory.

The program parses different configurations files with the order defined by the
UAPI group:
https://github.com/uapi-group/specifications/blob/main/specs/configuration_files_specification.md
and combines the parsed values to one single file.
Only drop-ins WITHOUT main configuration file will be merged:
https://github.com/uapi-group/specifications/blob/main/specs/configuration_files_specification.md#drop-ins-without-main-configuration-file

.SS "Example"
Files will be managed in following order:

* /usr/share/nsswitch.conf will not be regarded.
  Will not be regarded.

  Content:
  # comment for /usr/share/nsswitch.conf
  hosts: files
  
* /usr/share/nsswitch.conf.d/10-mdns.conf
  Vendor defined snipped

  Content:
  # comment for /usr/share/nsswitch.conf.d/10-mdns.conf
  hosts: mdns_minimal [NOTFOUND=return]

* /etc/nsswitch.conf.d/20-compat.conf
  Admin defined snipped

  Content:
  # comment for /etc/nsswitch.conf.d/20-compat.conf
  passwd:         compat
  group:          compat
  shadow:         compat

* /usr/share/nsswitch.conf.d/30-dns.conf
  Vendor defined snipped

  Content:
  # comment for /usr/share/nsswitch.conf.d/30-dns.conf
  hosts: dns

* /etc/nsswitch.conf.d/40-dns.conf
  Admin defined snipped

  Content:
  # comment for etc/nsswitch.conf.d/40-dns.conf (no double entry)
  hosts: dns

Generates /etc/nsswitch.conf with following content:

  # This file was generated by nsswitch-config. Do not edit it.
  # Changes can be done with single snippets, stored in <etc_dir>/nsswitch.conf.d/<snipped_name>
  # This file will not be overwritten if the comments above are removed.
  #
  # comment for /usr/share/nsswitch.conf.d/10-mdns.conf
  # comment for /usr/share/nsswitch.conf.d/30-dns.conf
  # comment for etc/nsswitch.conf.d/40-dns.conf (no double entry)
  hosts:mdns_minimal [NOTFOUND=return] dns
  # comment nr.1 for /etc/nsswitch.conf.d/20-compat.conf
  # comment nr.2 for /etc/nsswitch.conf.d/20-compat.conf
  passwd:compat
  group:compat
  shadow:compat

/etc/nsswitch.conf will be overwritten only, if it has not been created by
the admin. Otherwise it will be kept untouched and the results will be
written to /etc/nsswitch.conf.rpmnew.

.SH general OPTIONS
.TP
.B --vendordir <string>
  Directory for vendor specific configuration files.
  default: /usr/share

.TP
.B --etcdir <string>
  Directory for user changed configuration files (default: /etc).

.TP
.B --output <string>
  Path of generated nsswitch.conf file (default: /etc/nsswitch.conf).

.TP
.B --verbose
  Generates additional information.

.TP
.B --force
  Overwite output file although it has been created by the admin.

.TP
.B --is_generated
  Returns 0 if the output file has been generated by nsswitch-config.

.SH "SEE ALSO"
.PP 
nsswitch.conf(5)\&


NEWS
Version 0.0.7
* Replaced --admin_changed by --is_generated

Version 0.0.6
* --admin_changed: Showing more information.
* Fix: Crash while parsing empty values.

Version 0.0.5
* Create a /etc/nsswitch.conf.rpmnew if a admin created /etc/nsswitch.conf
  exists.
* New options:
  --force :
  Overwite output file although it has been created by the admin.
  --admin_changed :
  Check if the output file has been created by the admin.
* Templates for creating nsswitch.conf  

Version 0.0.4
* Added comment in nsswitch.conf that it has been created
  by nsswitch-config.
* There is only one backup file if it has not been generated
  by nsswitch-config.

Version 0.0.3
* Handle multiple versions of backup files.

Version 0.0.2
* Fixed memory leaks

Version 0.0.1
* Initial release


src/nsswitch-config.c
/*
  Copyright (C) 2025 SUSE LLC

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
*/

#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <pwd.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <libgen.h>
#include <dirent.h>
#include <stdlib.h>

#include "libeconf.h"
#include "libeconf_ext.h"

#define UNUSED(x) (void)(x)
#define PROGRAM_NAME "nsswitch-config"
#define COMMENT "#"
#define DELIMITERS ":"
#define VENDORDIR "/usr/share"
#define HEADERCOMMENT "This file was generated by nsswitch-config. Do not edit it."
#define NSSWITCHCONF "nsswitch.conf"
#define ETC "/etc"

/**
 * @brief Shows the usage.
 */
static void usage(void)
{
    fprintf(stderr, "Usage: %s [OPTIONS]\n\n", PROGRAM_NAME);
    fprintf(stderr, "\ngeneral Options:\n");
    fprintf(stderr, "--vendordir <string>: Directory for vendor specific configuration files (e.g. /usr/share).\n");
    fprintf(stderr, "                      default: %s\n", VENDORDIR);
    fprintf(stderr, "--etcdir <string>:    Directory for user changed configuration files (default: /etc).\n");    
    fprintf(stderr, "--output <string>:    Path of generated nsswitch.conf file (default: /etc/nsswitch.conf).\n");
    fprintf(stderr, "--is_generated:       Checks if the output file has been generated by nsswitch-config or by admin.\n");
    fprintf(stderr, "--force:              Overwrite output file regardless if it has been changed admin.\n");
    fprintf(stderr, "--verbose:            Generates additional information.\n");
}

/**
 * @brief printing error with linenr and filename
 */
static void print_error(const econf_err error)
{
  char *filename = NULL;
  uint64_t line_nr = 0;

  econf_errLocation( &filename, &line_nr);
  fprintf(stderr, "%s (line %d): %s\n", filename, (int) line_nr,
          econf_errString(error));
  free(filename);
}

/**
 * @brief Checking if the target file has already been generated
 *        by nsswitch-config.
 */
static int user_changed_file(const char *filename)
{
    if (access(filename, F_OK) == 0) {
        FILE *stream;
        char *line = NULL;
        size_t len = 0;
        ssize_t nread;
        int nsswitch_config_generated = 0;

        stream = fopen(filename, "r");
        if (stream == NULL) {
            fprintf(stderr, "Cannot open %s.\n", filename);
            return 0;
        }
        while ((nread = getline(&line, &len, stream)) != -1) {
            if (strstr(line, HEADERCOMMENT) != NULL) {
                nsswitch_config_generated = 1;
                break;
            }
        }
        free(line);
        fclose(stream);

        if (nsswitch_config_generated)
            return 0;
        else
            return 1;
    }
    return 0;
}

int main (int argc, char *argv[])
{
    econf_file *output_key_file = NULL;
    econf_file **key_file_list = NULL;
    size_t list_length = 0;
    econf_err econf_error = ECONF_SUCCESS;
    char *output_file = NULL;
    char *etc_main_file = NULL;
    char *usr_main_file = NULL;
    char *etc_dir = ETC;
    char *vendor_dir = VENDORDIR;
    int force = 0;
    char **keys = NULL;
    size_t key_count = 0;
    int ret = 0;
    econf_ext_value *ext_value = NULL;
    int verbose = 0;

    if (asprintf(&output_file, "%s/%s", ETC, NSSWITCHCONF) <0) {
        fprintf(stderr, "Cannot allocate memory.\n");
        return EXIT_FAILURE;
    }

    /* parse command line arguments. See getopt_long(3) */
    int opt;
    int index = 0;
    static struct option longopts[] = {
    /*   name,            arguments,      flag, value */
        {"help",          no_argument,       0, 'h'},
        {"vendordir",     required_argument, 0, 'v'},
        {"etcdir",        required_argument, 0, 'e'},
        {"output",        required_argument, 0, 'o'},
        {"verbose",       no_argument,       0, 'l'},
        {"force",         no_argument,       0, 'f'},
        {"is_generated",  no_argument,       0, 'a'},
        {0,               0,                 0,  0 }
    };

    while ((opt = getopt_long(argc, argv, "hvoelfa:", longopts, &index)) != -1) {
        switch(opt) {
        case 'v':
            vendor_dir = optarg;
            break;
        case 'e':
            etc_dir = optarg;
            break;
        case 'o':
            free (output_file);
            output_file = strdup(optarg);
            break;
        case 'l':
            verbose = 1;
            break;
        case 'f':
            force = 1;
            break;
        case 'a':
            if (!user_changed_file(output_file)) {
                fprintf(stderr, "true\n");
                return EXIT_SUCCESS;
            } else {
                fprintf(stderr, "false\n");
                return EXIT_FAILURE;
	    }
        case 'h':
            usage();
            return EXIT_SUCCESS;
        case '?':
        default:
            fprintf(stderr, "Try '%s --help' for more information.\n", PROGRAM_NAME);
            exit(EXIT_FAILURE);
            break;
        }
    }

    if (asprintf(&etc_main_file, "%s/%s", etc_dir, NSSWITCHCONF) <0) {
        fprintf(stderr, "Cannot allocate memory.\n");
        return EXIT_FAILURE;
    }

    if (asprintf(&usr_main_file, "%s/%s", vendor_dir, NSSWITCHCONF) <0) {
        fprintf(stderr, "Cannot allocate memory.\n");
        return EXIT_FAILURE;
    }

    econf_error = econf_newKeyFile( &output_key_file,
                                    ':', /* DELIMITERS */
                                    '#'  /* COMMENT */);
    if (econf_error) {
        print_error(econf_error);
        return EXIT_FAILURE;
    }    

    econf_error = econf_readDirsHistory( &key_file_list,
                                         &list_length,
                                         vendor_dir,
                                         etc_dir,
                                         "nsswitch",
                                         "conf",
                                         DELIMITERS,
                                         COMMENT);
    if (econf_error) {
        print_error(econf_error);
        ret = EXIT_FAILURE;
        goto cleanup;
    }

    /* sorting double entries */
    for (size_t i=0; i < list_length; i++) {
        for (size_t j = 0; j < list_length - i - 1; j++) {
            char *path1 = econf_getPath(key_file_list[j]);
            char *path2 = econf_getPath(key_file_list[j+1]);
            if (strcmp(basename(path1), basename(path2)) > 0) {
                econf_file * tmp = key_file_list[j];
                key_file_list[j] = key_file_list[j+1];
                key_file_list[j+1] = tmp;
            }
            free(path1);
            free(path2);
        }
    }

    for (size_t i=0; i < list_length; i++) {
        char *path = econf_getPath(key_file_list[i]);

        if (i < list_length - 1) {
            /* Checking if the next entry have the same basename */
            /* Ignore that one and take the next one             */
            char *next_path = econf_getPath(key_file_list[i+1]);
            if (strcmp(basename(path), basename(next_path)) == 0) {
                if (verbose)
                    fprintf(stderr, "Skipping: %s\n", path);
                continue;
            }
        }

        if (strcmp(path, output_file) == 0 ||
            strcmp(path, etc_main_file) == 0 ||
            strcmp(path, usr_main_file) == 0) {
            if (verbose)
                fprintf(stderr, "Skipping: %s\n", path);
            continue;
        }

        if (verbose)
            fprintf(stderr, "Evaluating: %s\n", path);

        free(path);

        econf_error = econf_getKeys(key_file_list[i],
                                    NULL, &key_count, &keys);
        if (econf_error && econf_error != ECONF_NOKEY) {
            print_error(econf_error);
            ret = EXIT_FAILURE;
            goto cleanup;
        }

        for (size_t k = 0; k < key_count; k++) {
            econf_ext_value *add_ext_value = NULL;
            char *value = NULL;
            char *add_value = NULL;

            econf_error = econf_getExtValue(key_file_list[i], NULL, keys[k], &add_ext_value);
            if (econf_error) {
                print_error(econf_error);
                econf_free(keys);
                ret = EXIT_FAILURE;
                goto cleanup;
            }
            econf_error = econf_getStringValue(key_file_list[i], NULL, keys[k], &add_value);
            if (econf_error) {
                print_error(econf_error);
                econf_freeExtValue(add_ext_value);
                econf_free(keys);
                ret = EXIT_FAILURE;
                goto cleanup;
            }

            /* adding value */
            econf_error = econf_getStringValue(output_key_file, NULL, keys[k], &value);
            if (econf_error == ECONF_NOKEY || value == NULL || strlen(value) == 0) {
                /* new entry */
                char *new_value = NULL;
                if (asprintf(&new_value, "%s%s", add_value != NULL ? " " : "", add_value != NULL ? add_value : "") <0) {
                      fprintf(stderr, "Cannot allocate memory.\n");
                      ret = EXIT_FAILURE;
                      free(value);
                      free(add_value);
                      econf_freeExtValue(add_ext_value);
                      econf_free(keys);
                      goto cleanup;
                }
                econf_error = econf_setStringValue(output_key_file, NULL, keys[k], new_value);
                free(new_value);
            } else if (econf_error == ECONF_SUCCESS && add_value != NULL && strlen(add_value) != 0) {
                /* appending entry */
                char *token, *str, *tofree;
                int found = 0;

                tofree = str = strdup(value);
                while ((token = strsep(&str, " \t\n")))
                        if (strcmp(token, add_value) == 0) found = 1;;
                free(tofree);

                if (found == 0) {
                   /* value is not in list */
                   char *new_value = NULL;
                   if (asprintf(&new_value, "%s %s", value, add_value) <0) {
                      fprintf(stderr, "Cannot allocate memory.\n");
                      ret = EXIT_FAILURE;
                      free(value);
                      free(add_value);
                      econf_freeExtValue(add_ext_value);
                      econf_free(keys);
                      goto cleanup;
                   }
                   econf_error = econf_setStringValue(output_key_file, NULL, keys[k], new_value);
                   free(new_value);
                }
            }
            free(value);
            free(add_value);
            if (econf_error != ECONF_SUCCESS) {
                print_error(econf_error);
                ret = EXIT_FAILURE;
                goto cleanup;
            }

            /* adding comments */
            if (econf_error == ECONF_SUCCESS &&
                add_ext_value->comment_before_key != NULL &&
                strlen(add_ext_value->comment_before_key) > 0) {
                if (econf_getExtValue(output_key_file, NULL, keys[k], &ext_value) == ECONF_SUCCESS) {
                    if (ext_value->comment_before_key == NULL) {
                        ext_value->comment_before_key = strdup(add_ext_value->comment_before_key);
                    } else {
                        char *new_comment = NULL;
                        if (asprintf(&new_comment, "%s\n%s", ext_value->comment_before_key, add_ext_value->comment_before_key) <0) {
                            ret = EXIT_FAILURE;
                            fprintf(stderr, "Cannot allocate memory.\n");
                            econf_freeExtValue(add_ext_value);
                            econf_freeExtValue(ext_value);
                            econf_free(keys);
                            goto cleanup;
                        } else {
                            free(ext_value->comment_before_key);
                            ext_value->comment_before_key = new_comment;
                        }
                    }
                    econf_error = econf_setExtValue(output_key_file, NULL, keys[k], ext_value);
                    econf_freeExtValue(ext_value);
                }
            }
            econf_freeExtValue(add_ext_value);
        }
        econf_free(key_file_list[i]);
        if (key_count > 0)
            econf_free(keys);
    }

    /* Writing comment that the output file has been generated by nsswitch-config */
    econf_error = econf_getKeys(output_key_file,
                                NULL, &key_count, &keys);
    if (econf_error && econf_error != ECONF_NOKEY) {
        print_error(econf_error);
        ret = EXIT_FAILURE;
        goto cleanup;
    }

    if (key_count > 1 &&
       econf_getExtValue(output_key_file, NULL, keys[0], &ext_value) == ECONF_SUCCESS) {

       if (ext_value->comment_before_key == NULL ||
           strstr(ext_value->comment_before_key, HEADERCOMMENT) == NULL) {

           char *new_comment = NULL;
           if (asprintf(&new_comment,
                        " %s\n Changes can be done with single snippets, stored in <etc_dir>/nsswitch.conf.d/"\
                        "<snipped_name>\n This file will not be overwritten if the comments above are removed.\n\n%s",
                        HEADERCOMMENT, ext_value->comment_before_key ? ext_value->comment_before_key : "") <0 ) {
              ret = EXIT_FAILURE;
              fprintf(stderr, "Cannot allocate memory.\n");
              econf_freeExtValue(ext_value);
              econf_free(keys);
              goto cleanup;
            } else {
               free(ext_value->comment_before_key);
               ext_value->comment_before_key = new_comment;
            }

            econf_error = econf_setExtValue(output_key_file, NULL, keys[0], ext_value);
            econf_freeExtValue(ext_value);
            if (econf_error) {
               print_error(econf_error);
               econf_free(keys);
               ret = EXIT_FAILURE;
               goto cleanup;
            }
        }
    }
    if (key_count > 1)
        econf_free(keys);

    if (user_changed_file(output_file) && !force) {
        char *dest = NULL;

        if (asprintf(&dest, "%s.rpmnew", output_file) <0 ) {
            fprintf(stderr, "Cannot allocate memory.\n");
            ret = EXIT_FAILURE;
            goto cleanup;
        }
        free (output_file);
        output_file = dest;
    }

    if (verbose) {
        fprintf(stderr, "Writing: %s\n", output_file);
    }

    econf_error = econf_writeFile(output_key_file, dirname(output_file), basename(output_file));
    if (econf_error) {
        print_error(econf_error);
        ret = EXIT_FAILURE;
    }    

 cleanup:
    free(key_file_list);
    econf_free(output_key_file);
    free (output_file);
    free (etc_main_file);
    free (usr_main_file);

    return ret;
}
